src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/CngKey.StandardProperties.cs

```diff
@@ -279,10 +296,19 @@ public IntPtr ParentWindowHandle
         {
             get
             {
-                string? provider = _providerHandle.GetPropertyAsString(ProviderPropertyName.Name, CngPropertyOptions.None);
-                if (provider == null)
-                    return null;
-                return new CngProvider(provider);
+                if (!_hasCachedProvider || _keyHandle.IsClosed)
```

Comment:
The fact that we need to check _keyHandle.IsClosed makes me think "obviously this should use a helper".  I think that helper would look like

```C#
private T GetCachedStringProperty<T>(ref bool hasValue, ref T field, string propertyName, Func<string?, T> resolver)
{
    if (!hasValue || _keyHandle.IsClosed)
    {
        string? value = _providerHandle.GetPropertyAsString(propertyName, CngPropertyOptions.None);
        T resolved = resolver(value);
        field = resolved;

        // Does the volatileness of the field flow through ref, or does this need to be explicit?
        Volatile.Write(ref hasValue, true);
    }

    return field;
}
```

and then the calls are

```C#
public CngProvider? Provider =>
    GetCachedStringProperty(ref _hasCachedProvider, ref _cachedProvider, ProviderPropertyName.Name, s => s is null ? null : new CngProvider(s));
```

which... is not necessarily better looking.

But the good news is it showed me a bug in this code :smile:

Reply:
That would only be usable for 2 out of the 3 properties being memoized, or we could use a bool field unnecessarily. Plus the fact that you had to ask yourself a question about volatile kind of puts in the "not worth it" camp.